### PR TITLE
add command, event, query and message middleware

### PR DESCRIPTION
**Pull request by sandrokeil**
This PR is ready for a review now. :tada: /cc @prolic @codeliner

### CODE REVIEW

**Comment by prolic** [comment]
I had only a short review so far. It looks very simple and intuitive. Good work!

File: src/Middleware.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Interface Middleware
+ *
+ * A middleware is a callable which handles the requested message type.
+ */
+interface Middleware
```
Comment (gabrielsch) [inline src/Middleware.php]: can't use [MiddlewareInterface](https://github.com/zendframework/zend-stratigility/blob/master/src/MiddlewareInterface.php) ?

File: docs/book/middleware.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+(`$request->getAttribute(\Prooph\Psr7Middleware\CommandMiddleware::NAME_ATTRIBUTE)`) called `prooph_command_name`. 
+This name is used for the `\Prooph\Common\Messaging\MessageFactory` to create the `\Prooph\Common\Messaging\Message` 
+object. The data for the command is extracted from the body of the request (`$request->getParsedBody()`) and must be an 
+array.
+
+## QueryMiddleware
+The `QueryMiddleware` dispatches the message data to the query bus system. This middleware needs an request attribute 
+(`$request->getAttribute(\Prooph\Psr7Middleware\QueryMiddleware::NAME_ATTRIBUTE)`) called `prooph_query_name`. 
+This name is used for the `\Prooph\Common\Messaging\MessageFactory` to create the `\Prooph\Common\Messaging\Message` 
+object. The data for the query is extracted from the body of the request (`$request->getParsedBody()`) and must be an
```
Comment (codeliner) [inline docs/book/middleware.md]: @prolic as far as I remember your ZF2-Query-API uses GET (query) params to populate query payload, right? I think we should do the same in the Query-Middleware.
Thoughts? @sandrokeil @prolic

**Comment by codeliner** [comment]
review done. Looks really great. Let's clarify the question in my comment and then it is ready to merge :+1:

File: docs/book/middleware.md (hunk 2 above)
Comment (prolic) [inline docs/book/middleware.md]: My Query-API uses GET to get the query name, filter parameters and pagination parameters and POST for the payload (additional information, needed for the query). This has some reasons:

1) It should be generic for my API, I don't want to have special controllers for every query
2) The filter parameters contains f.e. 'sort' this is used to sort the query result. On the other hand 'sort' can be a required payload parameter. So when I combine GET & POST I can have two different sort values with two different meanings. First (GET) is for the sorting result, second (POST) can be a required parameter for the query object (payload).

**Comment by sandrokeil** [comment]
@codeliner  Yes, using parsed body for the `QueryMiddleware` is wrong, but possible. Merging of query params and request params is not a good choice, see this example result:

```
result = {array} [8]
 action = "Prooph\Psr7Middleware\QueryMiddleware"
 email = "test@domain.com"
 name = "prooph"
 originalUri = {Zend\Diactoros\Uri} [9]
 prooph_query_name = "Prooph\ProophessorDo\Model\User\Query\UserList"
 REQUEST_METHOD = "GET"
 user_id = "f1a9d1c9-30eb-4352-82d6-4a899409f741"
 Zend\Expressive\Router\RouteResult = {Zend\Expressive\Router\RouteResult} [5]
```

What do you think about the solution to add a `prooph_attribute_list` request parameter, cycle through this list and add the parameter from the request to the payload? To support @prolic request body, we could add a additional attribute `prooph_use_request_body` and merge the payload with the parsed request body data. The request body overwrites all previous payload data, if defined.

Here is some pseudo code 

``` php
$attributeList = [];

foreach ($request->getAttribute(self::NAME_ATTRIBUTE_LIST) as $attribute)
{
    $attributeList[$attribute] = $request->getAttribute($attribute);
}

$query = $this->queryFactory->createMessageFromArray(
    $queryName,
    ['payload' => array_merge($request->getQueryParams(), $attributeList, $requestBody]
);
```

@prolic This middleware can not handle a query param _sort_ and a body param _sort_, because the body param overwrites the query param or is this ok in your case?

**Comment by prolic** [comment]
No it's not ok, and when you use $_POST and $_GET instead of $_REQUEST no
value gets overridden. You handle them separately and you can have separate
values. The psr-7 request can also distinguish between get and post.
Am 09.02.2016 09:51 schrieb "Sandro Keil" notifications@github.com:

**Comment by sandrokeil** [comment]
@prolic But how do you handle the values of $_POST and $_GET in the payload? I think you must use an own middleware for this behaviour and we only use the approach with query params and attribute list. For a REST API it would not work to use a POST to retrieve data. So we should use the HTTP specification behaviour (GET requests have no body).

**Comment by prolic** [comment]
Rest api with cqrs is weird anyway. As i said i use get for generic query
info, like query name, sort, filter, pagination and post for something like
user id param needed for the query.
If you won't change it i will use my own implementation as it's more
flexible.
Am 09.02.2016 10:09 schrieb "Sandro Keil" notifications@github.com:

**Comment by sandrokeil** [comment]
@prolic The default behaviour for the `QueryMiddleware` is to use query params now. However, I've added the support for POST HTTP requests and the parsed body data is stored under the key `data`.

Can we merge it now? /cc @codeliner

**Pull request merged**